File: app/src/main/java/com/android/architecture/example/SampleApplication.java
package com.android.architecture.example;

import android.app.Application;
import android.content.Context;

import timber.log.Timber;

public class SampleApplication extends Application{

    private SampleApplicationComponent component;

    @Override
    public void onCreate() {
        super.onCreate();

        component = DaggerSampleApplicationComponent.builder()
                .sampleApplicationModule(new SampleApplicationModule(this))
                .build();

        if (BuildConfig.DEBUG) {
            Timber.plant(new Timber.DebugTree());
        }
    }

    public SampleApplicationComponent getComponent() {
        return component;
    }

    public static SampleApplicationComponent component(Context context) {
        return ((SampleApplication) context.getApplicationContext()).getComponent();
    }
}

File: app/src/main/java/com/android/architecture/example/SampleApplication.kt
package com.android.architecture.example

import android.support.multidex.MultiDexApplication
import timber.log.Timber

open class SampleApplication: MultiDexApplication() {

    private val component: SampleApplicationComponent by lazy {
        DaggerSampleApplicationComponent.builder()
                .sampleApplicationModule(SampleApplicationModule(applicationContext, this))
                .build()
    }

    override fun onCreate() {
        super.onCreate()

        component().inject(this)

        if (BuildConfig.DEBUG) {
            Timber.plant(Timber.DebugTree())
        }

    }

    fun component(): SampleApplicationComponent {
        return component
    }

}